spring-integration-amqp/src/main/java/org/springframework/integration/amqp/outbound/RabbitStreamMessageHandler.java

```diff
@@ -0,0 +1,231 @@
+	 * String payload that contains json; the converter will set the content type to
+	 * {@code text/plain} which can be overridden to {@code application/json} by setting
+	 * the {@link AmqpHeaders#CONTENT_TYPE} message header. Default: false.
+	 * @param headersMappedLast true if headers are mapped after conversion.
+	 */
+	public void setHeadersMappedLast(boolean headersMappedLast) {
+		this.headersMappedLast = headersMappedLast;
+	}
+
+	/**
+	 * Return the {@link RabbitStreamOperations}.
+	 * @return the operations.
+	 */
+	public RabbitStreamOperations getStreamOperations() {
+		return this.streamOperations;
+	}
+
+	@Override
+	protected void handleMessageInternal(Message<?> requestMessage) {
+		CompletableFuture<Boolean> future;
+		com.rabbitmq.stream.Message streamMessage;
+		if (requestMessage.getPayload() instanceof com.rabbitmq.stream.Message) {
+			streamMessage = (com.rabbitmq.stream.Message) requestMessage.getPayload();
+		}
+		else {
+			MessageConverter converter = this.streamOperations.messageConverter();
+			org.springframework.amqp.core.Message amqpMessage = mapMessage(requestMessage, converter,
+					this.headerMapper, this.headersMappedLast);
+			streamMessage = this.streamOperations.streamMessageConverter().fromMessage(amqpMessage);
+		}
+		future = this.streamOperations.send(streamMessage);
+		handleConfirms(requestMessage, future);
+	}
+
+	private void handleConfirms(Message<?> message, CompletableFuture<Boolean> future) {
+		future.whenComplete((bool, ex) -> {
+			if (ex != null) {
+				this.failureCallback.failure(message, ex);
+			}
+			else {
+				this.successCallback.onSuccess(message);
+			}
+		});
+		if (this.sync) {
+			try {
+				future.get(this.confirmTimeout, TimeUnit.MILLISECONDS);
+			}
+			catch (InterruptedException ex) {
+				Thread.currentThread().interrupt();
+				throw new MessageHandlingException(message, ex);
+			}
+			catch (ExecutionException | TimeoutException ex) {
+				throw new MessageHandlingException(message, ex);
+			}
+		}
+	}
+
+	private static org.springframework.amqp.core.Message mapMessage(Message<?> message,
+			MessageConverter converter, AmqpHeaderMapper headerMapper, boolean headersMappedLast) {
+
+		MessageProperties amqpMessageProperties = new StreamMessageProperties();
+		return MappingUtils.mapMessage(message, converter, headerMapper, headersMappedLast, headersMappedLast,
+				amqpMessageProperties);
+	}
+
+	@Override
+	public void start() {
+	}
+
+	@Override
+	public void stop() {
+		this.streamOperations.close();
+	}
+
+	@Override
+	public boolean isRunning() {
+		return true;
+	}
+
+	/**
+	 * Callback for when publishing succeeds.
+	 */
+	public interface SuccessCallback {
+		/**
+		 * Called when the future completes with success.
+		 * Note that Exceptions raised by this method are ignored.
+		 * @param result the result of the future
+		 */
+		void onSuccess(Message<?> result);
+	}
+
+	/**
+	 * Callback for when publishing fails.
+	 */
+	public interface FailureCallback {
```

Comment:
Why is that?
Why not channels like we do in the `KafkaProducerMessageHandler`?

Reply:
I just copied what's in SCSt; but I agree.